### ISSUE TITLE
v1 roadmap

### ISSUE THREAD

**Issue by alecloudenback** [JuliaActuary/Yields.jl#75]
What else should be done before releasing a v1.0 version of the package? @kasperrisager @DimitarVanguelov

API to add:
- [x] #76
- [ ] #79

PRs to conclude on:
- [x] #56 

Other
- [x] Confirm okay with API of `Rates` additivity (per #84)

**Comment by DimitarVanguelov** [comment]
@alecloudenback, what is your expected timeline for 1.0? My only comment at this point, which is probably obvious, is that 1.0 should not be rushed since getting a package API right is crucial for usability (attracting users), long term maintenance and all that. That's not a comment on Yields' existing API (I am not fully familiar with it yet), just a general comment.

**Comment by kasperrisager** [comment]
I'd also put focus on API for a 1.0 (#56 is one direction to explore, sorry for being so slow on implementation). However, I also think it would be nice with a list of features that would give rough feature parity with comparable r/py/jl packages. Not because they necessarily have to go into a 1.0, but because the 1.0 API should make them easy to add. Interpolation is a good example of this: how can we make the API flexible enough that you can stuff any interpolation in, while slick enough to feel user friendly.

I've tried browsing for a good example of another package (.jl or other) that had a solid 1.0, maybe that's a road to explore a bit more.

Anyway, I think I'll try to pick #56 back up and think of other things while I'm at it

**Comment by alecloudenback** [comment]
I guess I should have outlined the rationale for tagging 1.0:

I largely agree with [Tamas here](https://discourse.julialang.org/t/how-can-we-encourage-julia-package-developers-to-release-version-1-0-0/29124/11), which is that:

1. once you start getting users and building dependecies, having the three digit semver provides more granular and meanful information to the downstream computers and users.
2. There's nothing super magical about 1.0. Going from `0.x` to `0.x+1` is effectively the same in pre-release `0.x.x`. In a lot of ways, there's more meaning to the whole semver scheme once you include the leading number.

Additionally:

3. I just see going `1.0` as a signal that the package's API has started to converge to a steady state. Users can learn the package without significant concern that they'll have to "re-learn" things.
4. Needing to go to `2.0` isn't a big deal at all IMO - Julia's Pkg and environment/dependency management means that users can easily control upgrades so I'm not super concerned about future breaking changes if they were to arise.

I've been using Yields for quite a while now and it's been a pretty stable API for a while, though admittedly in my activity I don't use the new SmithWilson-class curves.

The reason I started the issue was because aside from the things listed in the OP, I thought the remaining "issues" could likely be resolved without an API change.

**Comment by alecloudenback** [comment]
Another motivating factor for trying to resolve some of the last (currently) known API design decisions: In the coming months I'd like to write an introduction to JuliaActuay packages for one or more of the Society of Actuary's digital magazines. It would be nice to have a core package (Yields) be `1.0` for that

**Comment by alecloudenback** [comment]
> I've tried browsing for a good example of another package (.jl or other) that had a solid 1.0, maybe that's a road to explore a bit more.

@kasperrisager what comparable packages have you seen? 

Some I know of:
- Quantlib, the C++ library
  - I've tried to run/understand in the past but the code is dense and, well C++. And examples have always been light on exposition but heavy on the details.
- InterestRates.jl
  - (from the Yields.jl ReadMe): specializes in fast rate calculations aimed at valuing fixed income contracts, with business-day-level accuracy. Comparative comments: Yields.jl does not try to provide as precise controls over the timing, structure, and interpolation of the curve. Instead, Yields.jl provides a minimal, but flexible and intuitive interface for common modeling needs.

**Comment by kasperrisager** [comment]
@alecloudenback, haven't found good examples, apart from those you mention. I was also thinking about packages of similar complexity/maturity, but no luck yet finding one (or rather, not enough time spent on my side yet, I'm sure they are there).

InterestRates.jl seems to me to have a somewhat un-Julian API design with lots of unnecessary required methods on implementations of the abstract curve type. I think it's a great strength of the current Yields.jl API that it's nimble and flexible.

**Comment by alecloudenback** [comment]
I'm going to go ahead and tag `v1.0.0` not in staying that things are final (they aren't), but I think that it's:

- in a pretty good place
- signals a degree of stableness that I think has been achieved
- makes downstream compat easier

**Comment by alecloudenback** [comment]
I found this package in Python which has overlap in functionality between Yields.jl and ActuaryUtilities.jl: 
https://genedan.com/tmval/docs/intro.html
